README.md
[liburi](https://github.com/nevali/liburi) is a simple interface for parsing
URIs. Under the hood, the actual URI parsing is handled by
[uriparser](http://uriparser.sourceforge.net/) -- liburi aims to provide
an API which is easier to work with than uriparser's.

liburi provides:

* The ability to parse [IRIs](http://tools.ietf.org/html/rfc3987)
* Accessor methods for each parsed URI component and information about the URIs

liburi is licensed under the Apache License 2.0, while uriparser is licensed
under the [New BSD license](http://uriparser.git.sourceforge.net/git/gitweb.cgi?p=uriparser/uriparser;a=blob;f=COPYING).

### Building from git

You will need git, automake, autoconf and libtool. You'll also need a
DocBook 5 and DocBook-XSL toolchain (i.e., xsltproc and the stylesheets)
if you wish to rebuild the documentation. Both liburi and uriparser have
minimal external dependencies.

    $ git clone git://github.com/nevali/liburi.git
    $ cd liburi
    $ git submodule update --init --recursive
    $ autoreconf -i
    $ ./configure --prefix=/some/path
    $ make
    $ sudo make install


debian/changelog
liburi (1.0) stable; urgency=low

  * Initial release.

 -- Mo McRoberts <mo.mcroberts@bbc.co.uk>  Mon, 30 Jul 2014 14:14:01 +0000



t/match.c
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "p_testsuite.h"

static int testlen(const char *restrict file, const char *name, URI *restrict uri, struct urimatch *restrict test, size_t (*fn)(URI *restrict, char *restrict, size_t), size_t expected);

static int teststr(const char *restrict file, const char *name, URI *restrict uri, char *restrict buf, size_t buflen, struct urimatch *restrict test, size_t (*fn)(URI *restrict, char *restrict, size_t), const char *restrict expected);

int
test_urimatch(const char *file, struct urimatch *tests)
{
	URI *uri, *base;
	int c;
	int failed, didfail;
	char *buffer, *p;
	size_t bufsize, bl;
	int r;

	buffer = NULL;
	bufsize = 0;
	failed = 0;
	for(c = 0; tests[c].uri; c++)
	{
		base = NULL;
		bl = 0;
		if(tests[c].base)
		{
			base = uri_create_str(tests[c].base, NULL);
			bl = strlen(tests[c].base);
			if(!base)
			{
				fprintf(stderr, "%s: failed to construct base URI '%s': %s\n", file, tests[c].base, strerror(errno));
				failed++;
				continue;
			}			
		}
		bl += strlen(tests[c].uri);
		uri = uri_create_str(tests[c].uri, base);
		if(!uri)
		{
			fprintf(stderr, "%s: failed to construct URI '%s': %s\n", file, tests[c].uri, strerror(errno));
			uri_destroy(base);
			failed++;
			continue;
		}
		bl = (bl * 3) + 1;
		if(bl > bufsize)
		{
			p = (char *) realloc(buffer, bl);
			if(!p)
			{
				fprintf(stderr, "%s: failed to resize buffer from %u to %u bytes: %s\n", file, (unsigned) bufsize, (unsigned) bl, strerror(errno));
				exit(2);
			}
			buffer = p;
			bufsize = bl;
		}
		didfail = 0;
#define TESTLEN(name, mask) \
		if(tests[c].testmask & mask) \
		{ \
			didfail += testlen(file, #name, uri, &(tests[c]), uri_##name, tests[c].name##len); \
		}
		TESTLEN(scheme, UM_SCHEME_LEN);
		TESTLEN(auth, UM_AUTH_LEN);
		TESTLEN(host, UM_HOST_LEN);
		TESTLEN(port, UM_PORT_LEN);
		TESTLEN(path, UM_PATH_LEN);
		TESTLEN(query, UM_QUERY_LEN);
		TESTLEN(fragment, UM_FRAGMENT_LEN);
#undef TESTLEN
		if(tests[c].testmask & UM_RECOMPOSED_LEN)
		{
			didfail += testlen(file, "recomposed string", uri, &(tests[c]), uri_str, tests[c].recomposedlen);
		}
		if(tests[c].testmask & UM_ABSOLUTE)
		{
			if((r = uri_absolute(uri)) != tests[c].absolute)
			{
				didfail++;
				if(tests[c].absolute)
				{
					fprintf(stderr, "%s: URI expected to be absolute, but return value was %d for %s\n", file, r, tests[c].uri);
				}
				else
				{
					fprintf(stderr, "%s: URI expected to not be absolute, but return value was %d for %s\n", file, r, tests[c].uri);
				}
			}
		}
		if(tests[c].testmask & UM_PORT)
		{
			if((r = uri_portnum(uri)) != tests[c].portnum)
			{
				didfail++;
				fprintf(stderr, "%s: parsed port number (%d) does not match expected port number (%d) for URI %s\n", file, r, tests[c].portnum, tests[c].uri);
			}
		}
#define TESTSTR(name, mask) \
		if(tests[c].testmask & mask) \
		{ \
			didfail += teststr(file, #name, uri, buffer, bufsize, &(tests[c]), uri_##name, tests[c].name); \
		}
		TESTSTR(scheme, UM_SCHEME);
		TESTSTR(auth, UM_AUTH);
		TESTSTR(host, UM_HOST);
		TESTSTR(port, UM_PORT);
		TESTSTR(path, UM_PATH);
		TESTSTR(query, UM_QUERY);
		TESTSTR(fragment, UM_FRAGMENT);
#undef TESTSTR
		if(tests[c].testmask & UM_RECOMPOSED)
		{
			didfail += teststr(file, "recomposed string", uri, buffer, bufsize, &(tests[c]), uri_str, tests[c].recomposed);
		}
		if(didfail)
		{
			failed++;
		}
		uri_destroy(uri);
		uri_destroy(base);
	}
	free(buffer);
	return failed ? FAIL : PASS;
}

static int
testlen(const char *restrict file, const char *name, URI *restrict uri, struct urimatch *restrict test, size_t (*fn)(URI *restrict, char *restrict, size_t), size_t expected)
{
	size_t r;

	r = fn(uri, NULL, 0);
	if(r != expected)
	{
		fprintf(stderr, "%s: unexpected buffer size requested for %s (%u, expected %u) for URI '%s'\n", file, name, (unsigned) r, (unsigned) expected, test->uri);
		return 1;
	}
	return 0;
}

static int
teststr(const char *restrict file, const char *name, URI *restrict uri, char *restrict buffer, size_t buflen, struct urimatch *restrict test, size_t (*fn)(URI *restrict, char *restrict, size_t), const char *restrict expected)
{
	size_t r;

	memset(buffer, 0, buflen);
	r = fn(uri, buffer, buflen);
	if(r == (size_t) -1)
	{
		fprintf(stderr, "%s: failed to obtain %s for URI '%s'\n", file, name, test->uri);
		return 1;
	}
	if(r == 0 && expected)
	{
		fprintf(stderr, "%s: expected %s '%s', but result was NULL for URI '%s'\n", file, name, expected, test->uri);
		return 1;
	}
	if(r && !expected)
	{
		fprintf(stderr, "%s: expected NULL %s, but result was '%s' for URI '%s'\n", file, name, buffer, test->uri);
		return 1;
	}
	if(!expected)
	{
		return 0;
	}
	if(strcmp(buffer, expected))
	{
		fprintf(stderr, "%s: expected '%s' for %s, but result was '%s' for URI '%s'\n", file, expected, name, buffer, test->uri);
		return 1;
	}
	return 0;
}


t/testsuite.c
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "p_testsuite.h"

static struct test
{
	const char *name;
	int (*fn)(void);
	int expected;
} tests[] = {
	{ "test-parse-http-1.c", test_parse_http_1, PASS },

	{ NULL, NULL, 0 }
};

static const char *
passfail(int n)
{
	return n ? "FAIL" : "PASS";
}

int
main(int argc, char **argv)
{
	int c, r, unexpected, passed, failed;

	(void) argc;
	(void) argv;
	
	unexpected = 0;
	passed = 0;
	failed = 0;

	for(c = 0; tests[c].name; c++)
	{
		setlocale(LC_ALL, "C");
		printf("%40s ... ", tests[c].name);
		r = tests[c].fn();
		if(r == tests[c].expected)
		{
			printf("%s -- %d\n", passfail(r), r);
		}
		else
		{
			printf("%s -- %d (expected %s -- %d)\n", passfail(r), r, passfail(tests[c].expected), tests[c].expected);
			unexpected++;
		}
		if(r)
		{
			failed++;
		}
		else
		{
			passed++;
		}
	}
	printf("\n%c%c %d tests, %d passed, %d failed, %d unexpected result%c\n",
		   (unexpected ? '*' : '+'),  (unexpected ? '*' : '+'),
		   c, passed, failed, unexpected, unexpected == 1 ? ' ' : 's');
	return unexpected ? 1 : 0;
}



t/p_testsuite.h
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifndef P_TESTSUITE_H_
# define P_TESTSUITE_H_                 1

# include <stdio.h>
# include <stdlib.h>
# include <unistd.h>
# include <string.h>
# include <locale.h>
# include <errno.h>

# include "liburi.h"

# define PASS                           0
# define FAIL                           1

# define UM_NONE                        0
# define UM_SCHEME_LEN                  (1<<0)
# define UM_AUTH_LEN                    (1<<1)
# define UM_HOST_LEN                    (1<<2)
# define UM_PORT_LEN                    (1<<3)
# define UM_PATH_LEN                    (1<<4)
# define UM_QUERY_LEN                   (1<<5)
# define UM_FRAGMENT_LEN                (1<<6)
# define UM_RECOMPOSED_LEN              (1<<7)
# define UM_PORTNUM                     (1<<8)
# define UM_SCHEME                      (1<<9)
# define UM_AUTH                        (1<<10)
# define UM_HOST                        (1<<11)
# define UM_PORT                        (1<<12)
# define UM_PATH                        (1<<13)
# define UM_QUERY                       (1<<14)
# define UM_FRAGMENT                    (1<<15)
# define UM_RECOMPOSED                  (1<<16)
# define UM_ABSOLUTE                    (1<<17)

# define UM_ALL                         ((unsigned long) -1)

struct urimatch
{
	const char *uri;
	const char *base;
	const char *scheme;
	size_t schemelen;
	const char *auth;
	size_t authlen;
	const char *host;
	size_t hostlen;
	const char *port;
	size_t portlen;
	const char *path;
	size_t pathlen;
	const char *query;
	size_t querylen;
	const char *fragment;
	size_t fragmentlen;
	int portnum;
	int absolute;
	const char *recomposed;
	size_t recomposedlen;
	unsigned long testmask;
};

extern int test_urimatch(const char *file, struct urimatch *matches);

extern int test_parse_http_1(void);

#endif /*!P_TESTSUITE_H_*/


parse.c
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "p_liburi.h"

static size_t
uri_wctoutf8_(int *dest, wchar_t ch)
{
	if(ch < 0x7f)
	{
		dest[0] = ch;
		return 1;
	}
	if(ch < 0x07ff)
	{
		/* 110aaaaa 10bbbbbb */
		dest[0] = 0xc0 | ((ch & 0x0007c0) >>  6);
		dest[1] = 0x80 | (ch & 0x00003f);
		return 2;
	}
	if(ch < 0xffff)
	{
		/* 1110aaaa 10bbbbbb 10cccccc */
		dest[0] = 0xe0 | ((ch & 0x00f000) >> 12);
		dest[1] = 0x80 | ((ch & 0x000fc0) >> 6);
		dest[2] = 0x80 | (ch & 0x00003f);
		return 3;
	}
	/* 11110aaa 10bbbbbb 10cccccc 10dddddd */
	dest[0] = 0xf0 | ((ch & 0x1c0000) >> 18);
	dest[1] = 0x80 | ((ch & 0x03f000) >> 12);
	dest[2] = 0x80 | ((ch & 0x000fc0) >>  6);
	dest[3] = 0x80 | (ch & 0x00003f);
	return 4;
}

static size_t
uri_encode_wide_(char *dest, wchar_t ch)
{
	static const char hexdig[16] = {
		'0', '1', '2', '3', '4', '5', '6', '7',
		'8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
	};

	int utf8[6];
	size_t l, c;
	
	l = uri_wctoutf8_(utf8, ch);
	for(c = 0; c < l; c++)
	{
		*dest = '%';
		dest++;
		*dest = hexdig[utf8[c] >> 4];
		dest++;
		*dest = hexdig[utf8[c] & 15];
		dest++;
	}
	return l * 3;
}

/* Scan the URI string for wide characters and return the maximum storage
 * needed for their UTF-8 encoding
 */
static size_t
uri_widebytes_(const char *uristr, size_t nbytes)
{
	wchar_t ch;
	int r;
	const char *p;
	size_t numwide;

	mbtowc(&ch, NULL, 0);
	numwide = 0;
	for(p = uristr; *p;)
	{
		r = mbtowc(&ch, p, nbytes);
		if(r <= 0)
		{
			return (size_t) -1;
		}		
		if(ch < 33 || ch > 127)
		{
			/* Account for the full 6 bytes of UTF-8: we can't assume that
			 * the source string (and hence the return value of mbtowc()) is
			 * itself UTF-8, as it's locale-dependent.
			 */
			numwide += 6;
		}
		p += r;
	}
	return numwide;
}

/*
 * Map a potential IRI to a URI, see section 3.1 of RFC3987, converting
 * from locale-specific multibyte encoding to wide characters as we do
 */
static int
uri_preprocess_(char *restrict buf, const char *restrict uristr, size_t nbytes)
{
	wchar_t ch;
	char *bp;
	int r;

	mbtowc(&ch, NULL, 0);
	r = 0;
	for(bp = buf; nbytes && *uristr;)
	{
		r = mbtowc(&ch, uristr, nbytes);
		if(r <= 0)
		{
			return -1;
		}		
		if(ch < 33 || ch > 127)
		{			
			bp += uri_encode_wide_(bp, ch);
		}
		else
		{
			*bp = ch;
			bp++;
		}
		uristr += r;
		nbytes -= r;
	}
	*bp = 0;
	return 0;
}

URI *
uri_create_str(const char *restrict uristr, const URI *restrict base)
{
	UriParserStateA state;
	UriUriA absolute;
	URI *uri;
	size_t l, numwide;
	int r; 

	uri = (URI *) calloc(1, sizeof(URI));
	if(!uri)
	{
		return NULL;
	}
	l = strlen(uristr) + 1;
	numwide = uri_widebytes_(uristr, l);
	if(numwide == (size_t) -1)
	{
		uri_destroy(uri);
		return NULL;
	}
	uri->buf = (char *) malloc(l + numwide * 3);
	if(!uri->buf)
	{
		uri_destroy(uri);
		return NULL;
	}
	r = uri_preprocess_(uri->buf, uristr, l);
	if(r)
	{
		uri_destroy(uri);
		return NULL;
	}
	state.uri = &(uri->uri);
	if(uriParseUriA(&state, uri->buf) != URI_SUCCESS)
	{
		uri_destroy(uri);
		return NULL;
	}
	if(base)
	{
		if(uriAddBaseUriA(&absolute, &(uri->uri), &(base->uri)) != URI_SUCCESS)
		{
			uriFreeUriMembersA(&absolute);
			uri_destroy(uri);
			return NULL;
		}
		uriFreeUriMembersA(&(uri->uri));
		memcpy(&(uri->uri), &absolute, sizeof(UriUriA));
	}
	uriNormalizeSyntaxA(&(uri->uri));
	return uri;
}

URI *
uri_create_uri(const URI *restrict source, const URI *restrict base)
{
	URI *uri;
	UriParserStateA state;

	uri = (URI *) calloc(1, sizeof(URI));
	if(!uri)
	{
		return NULL;
	}
	if(base)
	{
		if(uriAddBaseUriA(&(uri->uri), &(source->uri), &(base->uri)) != URI_SUCCESS)
		{
			uri_destroy(uri);
			return NULL;
		}
	}
	else
	{
		uri->buf = strdup(source->buf);
		if(!uri->buf)
		{
			uri_destroy(uri);
			return NULL;
		}
		state.uri = &(uri->uri);
		if(uriParseUriA(&state, uri->buf) != URI_SUCCESS)
		{
			uri_destroy(uri);
			return NULL;
		}
	}
	uriNormalizeSyntaxA(&(uri->uri));
	return uri;
}

int
uri_destroy(URI *uri)
{
	if(uri)
	{
		uriFreeUriMembersA(&(uri->uri));
		free(uri->buf);
		free(uri);
	}
	return 0;
}	


props.c
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "p_liburi.h"

static ssize_t uri_get_(UriTextRangeA *restrict range, char *restrict buf, size_t bufsize);
static int uri_addch_(int ch, char *restrict *restrict buf, size_t *restrict buflen);

size_t
uri_scheme(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.scheme), buf, buflen);
}

size_t
uri_auth(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.userInfo), buf, buflen);
}

size_t
uri_host(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.hostText), buf, buflen);
}

size_t
uri_port(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.portText), buf, buflen);
}

size_t
uri_query(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.query), buf, buflen);
}

size_t
uri_fragment(URI *restrict uri, char *restrict buf, size_t buflen)
{
	return uri_get_(&(uri->uri.fragment), buf, buflen);
}

int
uri_absolute_path(URI *uri)
{
	if(uri->uri.absolutePath == URI_TRUE)
	{
		return 1;
	}
	if(uri->uri.hostText.first)
	{
		return 1;
	}
	if(uri->uri.scheme.first)
	{
		return 1;
	}
	return 0;
}

int
uri_absolute(URI *uri)
{
	if(uri->uri.scheme.first)
	{
		return 1;
	}
	return 0;
}

size_t
uri_path(URI *restrict uri, char *restrict buf, size_t buflen)
{
	size_t total, len;
	char *bp;
	UriPathSegmentA *p;
	
	if(!uri->uri.pathHead && !uri->uri.absolutePath)
	{
		if(buf && buflen)
		{
			*buf = 0;
		}
		return 0;
	}
	total = 0;
	bp = buf;
	if(uri_absolute_path(uri))
	{
		uri_addch_('/', &buf, &buflen);
		total++;
	}
	if(buf && buflen)
	{
		*buf = 0;
	}
	for(p = uri->uri.pathHead; p; p = p->next)
	{
		if(p != uri->uri.pathHead)
		{
			uri_addch_('/', &buf, &buflen);
			total++;
		}
		len = uri_get_(&(p->text), buf, buflen);
		if(len == (size_t) -1)
		{
			return (size_t) -1;
		}
		if(!len)
		{
			continue;
		}
		len--;
		total += len;
		if(buflen < len)
		{
			buflen = 0;
			buf = NULL;
		}
		else
		{
			if(buf)
			{
				buf += len;
			}
			buflen -= len;
		}
	}
	return total + 1;
}

int
uri_portnum(URI *uri)
{
	char buffer[32], *t;
	size_t len;
	unsigned long l;
	
	len = uri_port(uri, buffer, sizeof(buffer));
	if(len == (size_t) -1)
	{
		return -1;
	}
	if(len == 0 || len > sizeof(buffer) || !buffer[0])
	{
		return 0;
	}
	l = strtoul(buffer, &t, 10);
	if(t && *t)
	{
		return 0;
	}
	return (int) l;
}

URI_INFO *
uri_info(URI *uri)
{
	URI_INFO *p;
	char *buf;
	size_t buflen, r;

	buflen = 0;
#define getbuf(component) \
	r = uri_get_(&(uri->uri.component), NULL, 0);	  \
	if(r == (size_t) -1)							  \
	{												  \
		return NULL;								  \
	}												  \
	buflen += r;

	getbuf(scheme);
	getbuf(userInfo);
	getbuf(hostText);
	r = uri_path(uri, NULL, 0);
	if(r == (size_t) -1)
	{
		return NULL;
	}
	buflen += r;
	getbuf(query);
	getbuf(fragment);

#undef getbuf
	p = (URI_INFO *) calloc(1, sizeof(URI_INFO));
	if(!p)
	{
		return NULL;
	}
	buf = (char *) calloc(1, buflen);
	if(!buf)
	{
		free(p);
		return NULL;
	}
	p->internal = buf;
#define getbuf(component, member)						\
	r = uri_get_(&(uri->uri.component), buf, buflen);	\
	if(r)												\
	{													\
		p->member = buf;								\
		buf += r;										\
		buflen -= r;									\
	}

	getbuf(scheme, scheme);
	getbuf(userInfo, auth);
	getbuf(hostText, host);
	p->port = uri_portnum(uri);
	r = uri_path(uri, buf, buflen);
	if(r)
	{
		p->path = buf;
		buf += r;
		buflen -= r;
	}
	getbuf(query, query);
	getbuf(fragment, fragment);

#undef getbuf  
	return p;
}
	
/* Free a URI_INFO structure */
int
uri_info_destroy(URI_INFO *info)
{
	free(info->internal);
	free(info);
	return 0;
}

/* Compare two URIs and test for equality */
int
uri_equal(URI *a, URI *b)
{
	return uriEqualsUriA(&(a->uri), &(b->uri));
}

static ssize_t
uri_get_(UriTextRangeA *restrict range, char *restrict buf, size_t bufsize)
{
	size_t len, slen;

	if(!range->first)
	{
		if(buf && bufsize)
		{
			*buf = 0;
		}
		return 0;
	}
	if(range->afterLast)
	{
		len = range->afterLast - range->first;
	}
	else
	{
		len = strlen(range->first);
	}
	if(buf && bufsize)
	{
		/* Copy at most len or (bufsize - 1) characters, whichever
		 * is fewer.
		 */
		bufsize--;
		slen = bufsize < len ? bufsize : len;
		strncpy(buf, range->first, slen);
		buf[slen] = 0;
	}
	return len + 1;	
}

static int
uri_addch_(int ch, char *restrict *restrict buf, size_t *restrict buflen)
{
	if(*buf && *buflen)
	{
		**buf = ch;
		(*buf)++;
		**buf = 0;
		(*buflen)--;
	}
	return 1;
}


p_liburi.h
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifndef P_LIBURI_H_
# define P_LIBURI_H_                    1

# include <stdio.h>
# include <stdlib.h>
# include <wchar.h>
# include <errno.h>

# include "Uri.h"

# include "liburi.h"

struct uri_struct
{
	UriUriA uri;
	char *buf;
};

#endif /*!P_LIBURI_H_*/


util/uriparse.c
/*
 * Copyright 2012 Mo McRoberts.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <locale.h>
#include <errno.h>

#include "liburi.h"

static const char *short_program_name = "uriparse";
static const char *prefix = "";
static const char *uristr;
static const char *basestr;
static int verbose;
static int printuri;
static int omitempty;

static void parseargs(int argc, char **argv);
static void usage(void);
static URI *parseuris(void);
static int printcomp(URI *uri, const char *name, size_t (*fn)(URI *restrict, char *restrict, size_t), char **buffer, size_t *len);
static int printesc(const char *string);
static int print_uri(URI *uri);
static int print_components(URI *uri);

static void
parseargs(int argc, char **argv)
{
	const char *t;
	int ch;

	t = strrchr(argv[0], '/');
	if(t)
	{
		t++;
	}
	else
	{
		t = argv[0];
	}
	short_program_name = t;
	while((ch = getopt(argc, argv, "hvp:uo")) != -1)
	{
		switch(ch)
		{
		case 'h':
			usage();
			exit(EXIT_SUCCESS);
		case 'v':
			verbose = 1;
			break;
		case 'p':
			prefix = optarg;
			break;
		case 'u':
			printuri = 1;
			break;
		case 'o':
			omitempty = 1;
			break;
		default:
			usage();
			exit(EXIT_FAILURE);
		}
	}
	if(argc - optind < 1 || argc - optind > 2)
	{
		usage();
		exit(EXIT_FAILURE);
	}
	argc -= optind;
	argv += optind;
	uristr = argv[0];	
	if(argc >= 2)
	{
		basestr = argv[1];
	}
	else
	{
		basestr = NULL;
	}
}

static void
usage(void)
{
	fprintf(stderr, "Parse a URI and print its components\n\n");
	fprintf(stderr, "Usage: %s [OPTIONS] URI [BASE]\n\n", short_program_name);
	fprintf(stderr, "OPTIONS is one or more of:\n"
			"  -h                  Display this usage message and exit\n"
			"  -v                  Produce verbose output\n"
			"  -p PREFIX           Prefix output variable names with PREFIX\n"
			"  -u                  Print the parsed URI instead of components\n"
		    "  -o                  Omit printing components which are absent\n");
}

static URI *
parseuris(void)
{
	URI *uri, *rel, *base;

	if(basestr)
	{
		base = uri_create_str(basestr, NULL);		
		if(!base)
		{
			fprintf(stderr, "%s: failed to parse URI '%s': %s\n", short_program_name, basestr, strerror(errno));
			return NULL;
		}
		rel = uri_create_str(uristr, NULL);
		if(!rel)
		{
			fprintf(stderr, "%s: failed to parse URI '%s': %s\n", short_program_name, uristr, strerror(errno));
			return NULL;
		}
		uri = uri_create_uri(rel, base);
		if(!uri)
		{
			fprintf(stderr, "%s: failed to resolve '%s' against '%s': %s\n", short_program_name, uristr, basestr, strerror(errno));
		}
		uri_destroy(base);
		uri_destroy(rel);
		return uri;
	}
	uri = uri_create_str(uristr, NULL);
	if(!uri)
	{
		fprintf(stderr, "%s: failed to parse URI '%s': %s\n", short_program_name, uristr, strerror(errno));
		return NULL;
	}
	return uri;
}

static int
printcomp(URI *uri, const char *name, size_t (*fn)(URI *restrict, char *restrict, size_t), char **buffer, size_t *len)
{
	size_t r;
	char *p;

	r = fn(uri, *buffer, *len);
	if(r == (size_t) -1)
	{
		fprintf(stderr, "%s: failed to obtain %s: %s\n", short_program_name, name, strerror(errno));
		return -1;
	}
	if(r > *len)
	{
		p = (char *) realloc(*buffer, r);
		if(!p)
		{
			fprintf(stderr, "%s: failed to reallocate buffer from %lu to %lu bytes: %s\n", short_program_name, (unsigned long) *len, (unsigned long) r, strerror(errno));
			return -1;
		}
		*buffer = p;
		*len = r;
		r = fn(uri, *buffer, *len);
		if(r == (size_t) -1)
		{
			fprintf(stderr, "%s: failed to obtain %s: %s\n", short_program_name, name, strerror(errno));
			return -1;
		}
	}
	if(r == 0)
	{
		if(!omitempty)
		{
			printf("%s%s=''\n", prefix, name);
		}
		return 0;
	}	
	printf("%s%s=\"", prefix, name);
	printesc(*buffer);
	puts("\"");
	return 0;
}

static int
printesc(const char *string)
{
	const char *p;

	for(p = string; *p; p++)
	{
		switch(*p)
		{
		case '$':
		case '"':
		case '\\':
		case '`':
			putchar('\\');
			putchar(*p);
			break;
		default:
			if((unsigned char) *p < 32 || (unsigned char) *p > 127)
			{
				printf("\\%03o", *p);
				continue;
			}
			putchar(*p);
		}
	}
	return 0;
}

static int
print_uri(URI *uri)
{
	size_t len;
	char *buffer;

	len = uri_str(uri, NULL, 0);
	if(len == (size_t) -1)
	{
		fprintf(stderr, "%s: failed to recompose URI: %s\n", short_program_name, strerror(errno));
		return -1;
	}
	buffer = (char *) malloc(len);
	if(!buffer)
	{
		fprintf(stderr, "%s: failed to allocate %lu bytes: %s\n", short_program_name, (unsigned long) len, strerror(errno));
		return -1;
	}	
	len = uri_str(uri, buffer, len);
	if(len == (size_t) -1)
	{
		fprintf(stderr, "%s: failed to recompose URI: %s\n", short_program_name, strerror(errno));
		return -1;
	}
	puts(buffer);
	free(buffer);
	return 0;
}

static int
print_components(URI *uri)
{
	size_t len;
	char *buffer;
	int r;

	r = 0;
	len = 0;
	buffer = NULL;
	if(printcomp(uri, "scheme", uri_scheme, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "auth", uri_auth, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "host", uri_host, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "port", uri_port, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "path", uri_path, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "query", uri_query, &buffer, &len) == -1)
	{
		r = -1;
	}
	if(printcomp(uri, "fragment", uri_fragment, &buffer, &len) == -1)
	{
		r = -1;
	}
	free(buffer);

	return 0;
}

int
main(int argc, char **argv)
{
	URI *uri;
	int r;

	setlocale(LC_ALL, "");
	parseargs(argc, argv);
	uri = parseuris();
	if(!uri)
	{
		return 1;
	}
	if(printuri)
	{
		r = print_uri(uri);	
	}
	else
	{
		r = print_components(uri);
	}
	uri_destroy(uri);
	return (r ? 1 : 0);
}
